tools/pubsys/src/repo/check_expirations/mod.rs

```diff
@@ -0,0 +1,190 @@
+        );
+    }
+
+    expirations
+}
+
+fn check_expirations(
+    transport: &RepoTransport,
+    root_role_path: &PathBuf,
+    metadata_url: &Url,
+    targets_url: &Url,
+    expirations_between_now_and: Option<DateTime<Utc>>,
+) -> Result<()> {
+    // Create a temporary directory where the TUF client can store metadata
+    let workdir = tempdir().context(repo_error::TempDir)?;
+    let settings = Settings {
+        root: File::open(root_role_path).context(repo_error::File {
+            path: root_role_path,
+        })?,
+        datastore: workdir.path(),
+        metadata_base_url: metadata_url.as_str(),
+        targets_base_url: targets_url.as_str(),
+        limits: Limits::default(),
+        // We're gonna check the expiration ourselves
+        expiration_enforcement: ExpirationEnforcement::Unsafe,
+    };
+
+    // Load the repository
+    let repo = Repository::load(transport, settings).context(repo_error::RepoLoad {
+        metadata_base_url: metadata_url.clone(),
+    })?;
+    info!("Loaded TUF repo:\t{}", metadata_url);
+
+    info!("Root expiration:\t{}", repo.root().signed.expires);
+    info!("Snapshot expiration:\t{}", repo.snapshot().signed.expires);
+    info!("Targets expiration:\t{}", repo.targets().signed.expires);
+    info!("Timestamp expiration:\t{}", repo.timestamp().signed.expires);
+    // Check for upcoming metadata expirations if a timeframe is specified
+    if let Some(expiration_limit) = expirations_between_now_and {
+        let upcoming_expirations = find_upcoming_metadata_expiration(&repo, expiration_limit);
+        if !upcoming_expirations.is_empty() {
+            for (role, date) in upcoming_expirations {
+                if date < Utc::now() {
+                    error!("Repo '{}': '{}' expired on {}", metadata_url, role, date)
+                } else {
+                    warn!(
+                        "Repo '{}': '{}' expiring between now and {} on {}",
+                        metadata_url, role, expiration_limit, date
+                    )
+                }
+            }
+            return Err(Error::RepoExpirations {
+                metadata_url: metadata_url.clone(),
+            });
+        }
+    };
+
+    Ok(())
+}
+
+/// Common entrypoint from main()
+pub(crate) fn run(args: &Args, check_expirations_args: &CheckExpirationsArgs) -> Result<()> {
+    info!(
+        "Using infra config from path: {}",
+        args.infra_config_path.display()
+    );
+    let infra_config =
+        InfraConfig::from_path(&args.infra_config_path).context(repo_error::Config)?;
+    trace!("Parsed infra config: {:?}", infra_config);
+    let repo_config = infra_config
+        .repo
+        .as_ref()
+        .context(repo_error::MissingConfig {
+            missing: "repo section",
+        })?
+        .get(&check_expirations_args.repo)
+        .context(repo_error::MissingConfig {
+            missing: format!("definition for repo {}", &check_expirations_args.repo),
```

Comment:
I noticed a few other `context` calls change to `with_context` even though there's no copying/`format` inside, is there a benefit to that?  The goal of this suggestion was preventing a string build in the non-error case, since it was happening outside the closure.